src/surfaceintersection.jl

```diff
@@ -0,0 +1,111 @@
+surfacecrossing{T<:Real}(g::Taylor1{T}, g_old::Taylor1{T}, order::Int) = g[order+1]*g_old[order+1] < zero(T)
+surfacecrossing{T<:Real}(g::Taylor1{Taylor1{T}}, g_old::Taylor1{Taylor1{T}}, order::Int) = g[order+1][1]*g_old[order+1][1] < zero(T)
+surfacecrossing{T<:Real}(g::Taylor1{TaylorN{T}}, g_old::Taylor1{TaylorN{T}}, order::Int) = g[order+1][1][1]*g_old[order+1][1][1] < zero(T)
+
+function deriv{T<:Number}(n::Int, a::Taylor1{T})
+    @assert a.order ≥ n ≥ 0
+    if n==0
+        return a
+    elseif n==1
+        return derivative(a)
+    else
+        return deriv(n-1, derivative(a))
+    end
+end
+
+function taylorinteg{T<:Real, U<:Number}(f!, g, q0::Array{U,1}, t0::T, tmax::T,
+        order::Int, abstol::T; maxsteps::Int=500, nriter::Int=5, eventorder::Int=0)
+
+    # Allocation
+    const tv = Array{T}(maxsteps+1)
+    dof = length(q0)
+    const xv = Array{U}(dof, maxsteps+1)
+
+    # Initialize the vector of Taylor1 expansions
+    const t = Taylor1(T, order)
+    const x = Array{Taylor1{U}}(dof)
+    const dx = Array{Taylor1{U}}(dof)
+    const xaux = Array{Taylor1{U}}(dof)
+
+    # Initial conditions
+    @inbounds t[1] = t0
+    x0 = deepcopy(q0)
+    x .= Taylor1.(q0, order)
+    @inbounds tv[1] = t0
+    @inbounds xv[:,1] .= q0
+
+    #Some auxiliary arrays for root-finding/event detection/Poincaré surface of section evaluation
+    const g_val = zero(g(t,x,x))
+    const g_val_old = zero(g_val)
+    const slope = zero(U)
+    const dt_li = zero(U)
+    const dt_nr = zero(U)
+    const δt = zero(U)
+    const δt_old = zero(U)
+
+    const x_dx = vcat(x, dx)
+    const g_dg = vcat(g_val, g_val_old)
+    const x_dx_val = Array{U}( length(x_dx) )
+    const g_dg_val = vcat(evaluate(g_val), evaluate(g_val_old))
+
+    const tvS = Array{U}(maxsteps+1)
+    const xvS = similar(xv)
+    const gvS = similar(tvS)
+
+    #auxiliary range object for Newton-Raphson iterations
+    const nrinds = 1:nriter
+
+    # Integration
+    nsteps = 1
+    nevents = 1 #number of detected events
+    nextevord = eventorder+1
+    while t0 < tmax
+        δt_old = δt
+        δt = taylorstep!(f!, t, x, dx, xaux, t0, tmax, x0, order, abstol)
+        g_val = g(t, x, dx)
+        if surfacecrossing(g_val_old, g_val, eventorder)
```

Comment:
I understand this is intended as a _Poincaré Mapping_ detector. In this sense, it's not enough to just detect when does the solution `x(t)` crosses the surface `g`, but also it's important that the crossing is done in the same direction as `q0`. 
```julia
parallelcrossing{T<:Real}(x_last::Vector{T},x_old::Vector{T},q0::Vector{T}) = (x_last - x_old)⋅q0 > zero(T)
parallelcrossing{T<:Real}(x_last::Vector{Taylor1{T}},x_old::Vector{Taylor1{T}},q0::Vector{Taylor1{T}}) = ((x_last - x_old)⋅q0)[1] > zero(T)
parallelcrossing{T<:Real}(x_last::Vector{TaylorN{T}},x_old::Vector{TaylorN{T}},q0::Vector{TaylorN{T}}) = ((x_last - x_old)⋅q0)[1][1] > zero(T) 

poincarecrossing(g,g_old,order,x_last,x_old,q0) = surfacecrossing(g,g_old,order) && parallelcrossing(x_last,x_old,q0)
```
would do the job. Anyways, I don't know if it's intended for Poincaré Mappings or just for surface crossings, so we can discuss about this suggested change.

Reply:
Yes @blas-ko, my original idea was that these methods should be aimed towards constructing Poincaré maps, and I'm glad that they are helping you to construct them.

But now I think that there are cases when one wants to perform only "root-finding-while-integrating", or "detect events" defined by the roots of a given scalar function `g`, without constructing a Poincaré map. That's why now I think we should have some methods for plain, general "root-finding/event-detection", separated from specialized methods for Poincaré maps.

I agree, for Poincaré maps, we should only consider "positive crossings" in the sense that you mention. And also some other things must be taken into consideration, in particular, when we want to construct Poincaré maps with jet transport techniques. Besides what you mention, some other things should be considered:

- The times of crossing may, in general, be different for each point within the initial neighbourhood (bubble) of initial conditions defined by jet transport (current root-finding methods take care of this automatically).
- For conservative flows, only initial conditions within the initial bubble defined by jet transport,  with energy equal to `E0` (i.e., the energy of the reference initial condition `q0`) should be considered. This is easy to do for Hamiltonian flows with 2 degrees of freedom (as can be seen in the material of our JuliaCon talk for the Hénon-Heiles system), but as @lbenet has pointed (offline), this is non-trivial for more than 2 degrees of freedom.

Concretely, I would suggest to add a separate file, let's say `poincare.jl`, and add there any methods related specifically to Poincaré mapping, such as the ones you propose above. Moreover, I'd propose the name `poincare` for the analogous methods of `taylorinteg` for Poincaré maps. Please, feel free to submit a PR with these changes to the `jp/surface-intersection` branch from a fork!